src/rt/sched/behaviourcore.h

```diff
@@ -385,73 +385,95 @@ namespace verona::rt
         {
           std::get<0>(indexes[idx]) = i;
           std::get<1>(indexes[idx]) = &slots[j];
+          std::get<2>(indexes[idx]) = idx;
           idx++;
         }
       }
       auto compare = [](
-                       const std::tuple<size_t, Slot*> i,
-                       const std::tuple<size_t, Slot*> j) {
+                       const std::tuple<size_t, Slot*, size_t> i,
+                       const std::tuple<size_t, Slot*, size_t> j) {
 #ifdef USE_SYSTEMATIC_TESTING
-        return std::get<1>(i)->cown->id() > std::get<1>(j)->cown->id();
+        return std::get<1>(i)->cown->id() == std::get<1>(j)->cown->id() ?
+          std::get<2>(i) < std::get<2>(j) :
+          std::get<1>(i)->cown->id() < std::get<1>(j)->cown->id();
 #else
-        return std::get<1>(i)->cown > std::get<1>(j)->cown;
+        return std::get<1>(i)->cown == std::get<1>(j)->cown ?
+          std::get<2>(i) < std::get<2>(j) :
+          std::get<1>(i)->cown < std::get<1>(j)->cown;
 #endif
       };
 
       if (count > 1)
         std::sort(indexes.get(), indexes.get() + count, compare);
 
       // First phase - Acquire phase.
-      Cown *prev_cown = nullptr;
-      for (size_t i = 0; i < count; i++)
+      size_t i = 0;
+      while (i < count)
       {
         auto cown = std::get<1>(indexes[i])->cown;
         auto body = bodies[std::get<0>(indexes[i])];
-        auto prev = cown->last_slot.exchange(
-          std::get<1>(indexes[i]), std::memory_order_acq_rel);
-        auto* ec_ptr = &ec[std::get<0>(indexes[i])];
+        auto last_slot = std::get<1>(indexes[i]);
+        auto first_body = body;
+        size_t ii = i;
+        while (i < count - 1)
+        {
+          auto cown_next = std::get<1>(indexes[i + 1])->cown;
+          if (cown_next != cown)
+            break;
 
-        yield();
+          body = bodies[std::get<0>(indexes[i + 1])];
+          last_slot->set_behaviour(body);
+
+          last_slot = std::get<1>(indexes[i + 1]);
+          // cown = cown_next;
+          i++;
+        }
+        i++;
 
+        auto prev =
+          cown->last_slot.exchange(last_slot, std::memory_order_acq_rel);
+
+        // set_behaviour to the fist_slot
+        yield();
         if (prev == nullptr)
         {
+          // this is wrong - should only do it for the last one
           Logging::cout() << "Acquired cown: " << cown << " for behaviour "
                           << body << Logging::endl;
-          (*ec_ptr)++;
+
+          ec[std::get<0>(indexes[ii])]++;
+
           if (transfer == NoTransfer)
```

Comment:
I think something goes wrong here with `YesTransfer` if their are duplicate cowns, and the cown is not currently held.  I think the client will be passing multiple RC to the same object, and then it needs to remove them, e.g.

```
when (a) { .. } + when (a) { .. }
```

with YesTransfer, and the underlying `a` is not held, will leak one RC to the cown `a`.

Reply:
Does it matter that it passes multiple behaviours? In the case of having the same cown multiple times, this will only happen once. Are you suggesting it should happen multiple times as the number of behaviours enqueued on the same cown?

Reply:
Let me make some updates and get back to you on this one because indeed RC is broken as is right now.

But if I read what was there before correctly, I see the following:
if prev == nullptr && transfer == NoTransfer then acquire
if prev != nullptr && transfer == YesTransfer then release

When we have a chain of behaviours for the same cown, the value of prev matters only for the first behaviour in the chain. For all the other behaviours, "prev" will not be nullptr, since there are behaviours there before.

Do I read this correctly?

If yes, this means that if NoTransfer and prev == nullptr we need to acquire once. If YesTransfer, then we need to release for at least chain_length - 1.

Does this make sense?